tests/integration/test_charm.py

```diff
@@ -0,0 +1,218 @@
+from inspect import cleandoc
+from pathlib import Path
+from typing import TYPE_CHECKING, cast
+
+import pytest
+import yaml
+from pytest_operator.plugin import OpsTest
+
+import charm
+
+if TYPE_CHECKING:
+    from juju.action import Action
+    from juju.application import Application
+    from juju.model import Model
+    from juju.unit import Unit
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path('./charmcraft.yaml').read_text())
+APP_NAME = METADATA['name']
+SENTINEL_PATH = '/books/sentinel'
+KNOWN_LIBRARY_INFO_OUTPUT: dict[Path, dict[charm.LibraryInfoFormat, str]] = {
+    Path('files/library.zip'): {
+        'ls-1': cleandoc(
+            """
+            John Schember
+            metadata.db
+            """
+        ),
+        'tree': cleandoc(
+            """
+            .
+            ├── John Schember
+            │   └── Quick Start Guide (1)
+            │       ├── cover.jpg
+            │       ├── metadata.opf
+            │       └── Quick Start Guide - John Schember.epub
+            └── metadata.db
+
+            2 directories, 4 files
+            """
+        ),
+    },
+    Path('tests/library-austen-flat.zip'): {
+        'ls-1': cleandoc(
+            """
+            Jane Austen
+            John Schember
+            metadata.db
+            """
+        ),
+        'tree': cleandoc(
+            """
+            .
+            ├── Jane Austen
+            │   └── Pride and Prejudice (2)
+            │       ├── cover.jpg
+            │       ├── metadata.opf
+            │       └── Pride and Prejudice - Jane Austen.epub
+            ├── John Schember
+            │   └── Quick Start Guide (1)
+            │       ├── cover.jpg
+            │       ├── metadata.opf
+            │       └── Quick Start Guide - John Schember.epub
+            └── metadata.db
+
+            4 directories, 7 files
+            """
+        ),
+    },
+}
+KNOWN_LIBRARY_INFO_OUTPUT.update(
+    {
+        Path('tests/library-austen-nested.zip'): KNOWN_LIBRARY_INFO_OUTPUT[
+            Path('tests/library-austen-flat.zip')
+        ]
+    }
+)
+
+
+@pytest.mark.skip_if_deployed
+@pytest.mark.abort_on_fail
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build and deploy calibre-web charm, attaching resources and storage.
+
+    Fails sometimes on my machine due to failure to attach storage, but passes in a fresh VM.
+    """
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm('.')
+    # deploy and attach storage
+    model = cast('Model', ops_test.model)
+    resources = {
+        'calibre-web-image': METADATA['resources']['calibre-web-image']['upstream-source'],
+        'calibre-library': './empty.zip',
+    }
+    await model.deploy(charm, resources=resources, application_name=APP_NAME)
+    await model.wait_for_idle(apps=[APP_NAME], status='active', raise_on_blocked=True, timeout=600)
+
+
+async def test_status_for_set_config(ops_test: OpsTest):
```

Comment:
My inclination would be for this to be a parametrized test, but it's fine as-is too.

Reply:
Good call, thanks